Cleanup temporary files.

dargor opened the pull request:
Closes #88

jkroepke commented:
Hi @dargor!

Good catch! Could you add a test case for this?

dargor commented:
I am trying but I am not sure how to catch a randomly generated tmpfile in the tests.

The tests themselves use and leave quite a bunch of tmpfiles/tmpdirs too, I will make a new PR for this.

jkroepke commented:
Hi,

after #90 is merged, I would understand the amount of tmp files is reduced.

Otherwise, would setting `TMPDIR` (See https://linux.die.net/man/1/mktemp for mktemp behavior) a solution here to define the location of the files generated by the vault driver? + checking if there are some leftovers?

dargor commented:
I rebased locally and found `kubeval.bats`, `lint.bats`, `template.bats` and `vault.bats` guilty of leaking empty (aka 0 bytes) files. I am currently checking if the tests or the drivers are to blame for that, but at least the vault driver do not leak secrets anymore with this PR.

Overriding `TMPDIR` could work, I will try that.

dargor commented:
I found where most of these 0 bytes files came from : scripts/commands/helm.sh had a `-s` test and removed only non empty files. I changed it to remove tmpfiles unconditionally.

Now I only have `vault.bats` still leaking files.

jkroepke commented on scripts/drivers/_custom.sh:
```diff
@@ -42,6 +42,7 @@ driver_decrypt_file() {
         _SECRET="${EXPRESSION#* }"
 
         if ! SECRET=$(_custom_driver_get_secret "${type}" "${_SECRET}"); then
+            rm -f "${output_yaml}" "${output_yaml_anchors}"
```
I would recommend to use trap function here.

dargor replied:
There is already a trap setup elsewhere, when I use a new trap it overwrite the previous one.

jkroepke commented:
I looked though the code again and figure out that i'm set an own `TMPDIR` as base for all temporary files.

https://github.com/jkroepke/helm-secrets/blob/f1820465e119ae90d30d1fc39a8f6d137055bd86/scripts/run.sh#L13

This temp directory is removed in a trap function which should be called at the end of the script execution.

https://github.com/jkroepke/helm-secrets/blob/f1820465e119ae90d30d1fc39a8f6d137055bd86/scripts/lib/common.sh#L47-L51

It means each file generated by `mktemp` should be delete after the script finished. That should include

https://github.com/jkroepke/helm-secrets/blob/f1820465e119ae90d30d1fc39a8f6d137055bd86/scripts/drivers/_custom.sh#L32-L33

, too.

I would like to know, whats the root cause of #88 and why the trap function not works.

dargor commented:
Sounds like a missing `EXPORT` to me, like
```diff
diff --git a/scripts/run.sh b/scripts/run.sh
index 56a684c..5e9680e 100755
--- a/scripts/run.sh
+++ b/scripts/run.sh
@@ -11,6 +11,7 @@ SCRIPT_DIR="$(dirname "$0")"

 # Create temporary directory
 TMPDIR="${HELM_SECRETS_DEC_TMP_DIR:-$(mktemp -d)}"
+export TMPDIR

 # Output debug infos
 QUIET="${HELM_SECRETS_QUIET:-false}"
```

When I revert the `rm` I added and add only the `EXPORT` I no longer have leaks, but some tests errors out (and works when I comment the `EXPORT` back) : `kubeval`, `lint` and `template`.

jkroepke commented:
I follow up the `export TMPDIR` in #93 .

Additionally, I figure out that on Mac OS does not respect `TMPDIR`. See: https://unix.stackexchange.com/a/555214 

Since I'm running Mac OS locally, I could add an proposal to resolve this topic on MacOS, too.
